### PR TITLE
overlord/state: introduce Status/SetStatus and Progress/SetProgress on Task

### PR DESCRIPTION
**Pull request by pedronis**
This introduces Status/SetStatus and Progress/SetProgress on Task and the Status type itself.

### CODE REVIEW

**Comment by pedronis** [comment]
requires #578

File: overlord/state/change.go [hunk 1]
```diff
@@ -23,6 +23,17 @@ import (
 	"encoding/json"
 )
 
+// Status is used for status values for changes and tasks.
+type Status int
+
+// Admitted status values for changes and tasks.
+const (
+	Running Status = 0
```
Comment (zyga) [inline overlord/state/change.go]: perhaps the perfect spot to use iota unless we specifically want to use particular values forever across re-shuffling of those constants.

File: overlord/state/task.go [hunk 2]
```diff
@@ -46,20 +48,26 @@ func newTask(state *State, id, kind, summary string) *Task {
...
 type marshalledTask struct {
-	ID      string                      `json:"id"`
-	Kind    string                      `json:"kind"`
-	Summary string                      `json:"summary"`
-	Data    map[string]*json.RawMessage `json:"data"`
+	ID          string                      `json:"id"`
+	Kind        string                      `json:"kind"`
+	Summary     string                      `json:"summary"`
+	Status      Status                      `json:"status"`
+	CurProgress int                         `json:"cur_progess"`
```
Comment (zyga) [inline overlord/state/task.go]: ``current-progress`` or ``cur-progress`` per https://trello.com/c/wXuctexM/1-dashes-everywhere

File: overlord/state/task.go [hunk 3]
```diff
@@ -46,20 +48,26 @@ func newTask(state *State, id, kind, summary string) *Task {
...
+	TotProgress int                         `json:"tot_progress"`
```
Comment (zyga) [inline overlord/state/task.go]: ``total-progress`` or ``tot-progress`` per https://trello.com/c/wXuctexM/1-dashes-everywhere

**Comment by zyga** [comment]
Just suggestion to tweak the total progress to max progress. Overall +1.

File: overlord/state/change.go (hunk 1 above)
Comment (niemeyer) [inline overlord/state/change.go]: Works either way, but we _definitely_ don't want to change those values. That's being serialized into state!

File: overlord/state/task.go [hunk 4]
```diff
@@ -28,11 +28,13 @@ import (
...
-	id      string
-	kind    string
-	summary string
-	data    customData
+	state                    *State
+	id                       string
+	kind                     string
+	summary                  string
+	status                   Status
+	curProgress, totProgress int
```
Comment (niemeyer) [inline overlord/state/task.go]: Break down into two lines for a more pleasant formatting?

File: overlord/state/task.go (hunk 2 above)
Comment (niemeyer) [inline overlord/state/task.go]: Yes, I'd go with one of these:

* `"progress": {"current": 0, "total", 1}`
* `current-progress` and `total-progress`
* `current` and `total`

Comment (niemeyer) [inline overlord/state/task.go]: Probably the first.. gives us a value to pass around by itself.

File: overlord/state/task.go (hunk 3 above)
Comment (niemeyer) [inline overlord/state/task.go]: Hint: _max_ is not an actual word. :-)

File: overlord/state/task.go (hunk 2 above)
Comment (niemeyer) [inline overlord/state/task.go]: And gives us progress.Total, which is much nice. Yeah, I'd go with that.

File: overlord/state/task.go [hunk 5]
```diff
@@ -95,6 +106,41 @@ func (t *Task) Summary() string {
...
+	t.state.ensureLocked()
+	t.status = s
+}
+
+// Progress returns the current progress for the task.
+// If progress is not explicitly set, it returns (0, 1) if the status is
+// Running or Waiting and (1, 1) otherwise.
+func (t *Task) Progress() (cur, total int) {
+	t.state.ensureLocked()
+	if t.curProgress == 0 && t.totProgress == 0 {
```
Comment (niemeyer) [inline overlord/state/task.go]: We can test the total value only here. It'd be an error to have cur > 0 if total is 0, so we can fallback to the default behavior.

File: overlord/state/task.go [hunk 6]
```diff
@@ -28,11 +28,14 @@ import (
...
-	summary string
-	data    customData
+	state                    *State
+	id                       string
+	kind                     string
+	summary                  string
+	status                   Status
+	curProgress              int
+	totProgress              int
+	data                     customData
```
Comment (niemeyer) [inline overlord/state/task.go]: go fmt

**Comment by niemeyer** [comment]
Some simple tweaks, and LGTM!

**Pull request merged**